### PR TITLE
Task Completion / Database Refactor

### PR DESCRIPTION
**Pull request by adventurist**
Tasks need more nuanced completion states than simply `complete` and `incomplete`

We need to be able to track ones which have failed, give them a chance to run once more, and then separate tasks which have failed upon retry.

We also needed to refactor the database layer to:
- Allow for multiple filters of particular types (BETWEEN, WHERE x in (<list>))
- replace all the overloads with templates.

### CODE REVIEW

File: include/executor/scheduler.hpp [hunk 1]
```diff
@@ -12,10 +11,33 @@
...
+ * SCHEDULED  - Has not yet run
+ * COMPLETED  - Ran successfully
+ * FAILED     - Ran once and failed
+ * RETRY_FAIL - Failed on retry
+ */
+namespace Completed {
+  static constexpr int SCHEDULED = 0;
+  static constexpr int SUCCESS = 1;
+  static constexpr int FAILED = 2;
+  static constexpr int RETRY_FAIL = 3;
```
Comment (lovelyinyoung) [inline include/executor/scheduler.hpp:32]: Do we also need "Retry_success"?

File: src/database_connection.cpp [hunk 2]
```diff
@@ -79,16 +80,74 @@ std::string updateStatement(UpdateReturnQuery query, std::string returning, bool
...
+template <typename FilterA, typename FilterB>
+std::string getVariantFilterStatement(std::vector<std::variant<FilterA, FilterB>> filters) {
+  std::string filter_string{};
+  auto idx = 0;
+  auto filter_count = filters.size();
+  for (const auto& filter : filters) {
+    if (filter.index() == 0) {
+      filter_string += filterStatement(std::get<0>(filter));
+    } else {
+      filter_string += filterStatement(std::get<1>(filter));
```
Comment (lovelyinyoung) [inline src/database_connection.cpp:113]: can there be more than 1 filter index?

File: src/database_connection.cpp [hunk 3]
```diff
@@ -318,6 +378,27 @@ QueryResult DatabaseConnection::query(MultiFilterSelect query) {
...
 
+QueryResult DatabaseConnection::query(MultiVariantFilterSelect query) {
+  pqxx::result pqxx_result = performSelect(query);
+  QueryResult result{};
+  result.table = query.table;
+  auto count = query.fields.size();
+
+  for (auto row : pqxx_result) {
+    int index = 0;
+    for (const auto &field : row) {
```
Comment (lovelyinyoung) [inline src/database_connection.cpp:389]: in c++, is there for (int i = 0; i < object.length(); i++) { // function }
It looks weird to have index outside of for loop and it's used inside like that.

File: include/executor/scheduler.hpp (hunk 1 above)
Comment (adventurist) [reply include/executor/scheduler.hpp:32]: We could use it for more nuanced record keeping, but I'm not sure if it's necessary. We more or less just needed to differentiate between tasks which may have failed for edge case scenarios vs tasks which most probably fail on every retry.

File: src/database_connection.cpp (hunk 2 above)
Comment (adventurist) [reply src/database_connection.cpp:113]: The filter index range is dictated by the number of template arguments being given to `std::variant` on line 105

File: src/database_connection.cpp (hunk 3 above)
Comment (adventurist) [reply src/database_connection.cpp:389]: There is.. I was half way convinced to use a plain for loop for that reason.

**Comment by adventurist** [comment]
Made more changes, especially the removal of many definitions of Database::query, which has been replaced by a templated method

File: src/database_connection.cpp [hunk 4]
```diff
@@ -247,97 +305,41 @@ QueryResult DatabaseConnection::query(DatabaseQuery query) {
     }
     case QueryType::SELECT: {
       pqxx::result pqxx_result = performSelect(query);
-      QueryResult result{};
-      result.table = query.table;
-      auto count = query.fields.size();
-
-      for (auto row : pqxx_result) {
+      QueryResult result{.table=query.table};
```
Comment (adventurist) [inline src/database_connection.cpp:308]: It's really around here where the relevant changes were made (removing explicit method definitions and replacing with template method)

**Pull request merged**